File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/observer/solution/Participant.java
// Participant.cpp

public abstract class Participant
{
	public Participant()
	{
		myMeetings = new DpVector<Meeting>();
	}

	public void add(Meeting aMeeting)
	{
		if (aMeeting == null)
		{
			throw new DpException("void Participant.add(Meeting aMeeting) -- Invalid aMeeting.");
		}

		System.out.println(
			  "Participant.add("
			+ aMeeting.toString()
			+ ")"
			);
		myMeetings.add(aMeeting);
	}

	public void remove(Meeting aMeeting)
	{
		if (aMeeting == null)
		{
			throw new DpException("void remove(Meeting aMeeting) -- Invalid aMeeting.");
		}

		System.out.println(
			  "Participant.remove("
			+ aMeeting.toString()
			+ ")"
			);
		// I was too lazy to implement this. :-(
	}

	public void update(Meeting aMeeting)
	{
		if (aMeeting == null)
		{
			throw new DpException("void update(Meeting aMeeting) -- Invalid aMeeting.");
		}

		System.out.println(
			  "Participant.update("
			+ aMeeting.toString()
			+ ")"
			);
	}

	private DpVector<Meeting> myMeetings;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/observer/solution/Participant.m
% Participant.m

classdef (Abstract) Participant < handle
	properties (GetAccess = private, SetAccess = private)
		Meetings % DpVector of Meeting
	end

	methods
		% constructor:
		function theParticipant = Participant
			theParticipant.Meetings = DpVector;
		end

		function add(theParticipant, aMeeting)
			if isempty(aMeeting)
				error('Invalid aMeeting.');
			end

			fprintf('Participant.add(%s)\n', aMeeting.toString);
			theParticipant.Meetings.add(aMeeting);
		end

		function remove(theParticipant, aMeeting)
			if isempty(aMeeting)
				error('Invalid aMeeting.');
			end

			fprintf('Participant.remove(%s)\n', aMeeting.toString);
			% I was too lazy to implement this. :-(
		end

		function update(theParticipant, aMeeting)
			if isempty(aMeeting)
				error('Invalid aMeeting.');
			end

			fprintf('Participant.update(%s)\n', aMeeting.toString);
		end
	end
end
